swagger/smart_transfer_protocol_ws.yaml

```diff
@@ -0,0 +1,421 @@
+            dataSlices:
+              type: array
+              description: List of data slice file hashes referenced by the new blocks
+              items:
+                type: string
+                format: multihash
+            checkpoints:
+              type: array
+              description: List of checkpoint file hashes referenced by the new blocks
+              items:
+                type: string
+                format: multihash
+  
+  dataset/pull/data/negotiate:
+    subscribe:
+      operationId: odf.dataset.pull.data.negotiate
+      summary: Requests to negotiate data download protocol for 1 or more files
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPullDataNegotiate
+        name: DatasetPullDataNegotiate
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - datasetId
+            - dataFiles
+          properties:
+            datasetId:
+              type: string
+              format: did
+              description: Dataset identifier     
+            dataFiles:
+              type: array
+              description: List of data file references in this bulk
+              items:
+                $ref: '#/components/schemas/DataFileReference'
+      
+  dataset/pull/data/negotiate/response:
+    publish:
+      operationId: odf.dataset.pull.data.negotiate.response
+      summary: Response on negotiated status of data operations during pull
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPullDataNegotiateResponse
+        name: DatasetPullDataNegotiateResponse
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - datasetId
+            - negotiatedDownloads
+          properties:
+            datasetId:
+              type: string
+              format: did
+              description: Dataset identifier        
+            negotiatedDownloads:
+              type: array
+              description: Negotiated download operations for requested bulk of files
+              items:
+                $ref: '#/components/schemas/NegotiatedDownload'
+
+  dataset/push/request:
+    subscribe:
+      operationId: odf.dataset.push.request
+      summary: Initiates dataset push to the ODF server
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPushRequest
+        name: DatasetPushRequest
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - datasetId
+            - knownHead
+            - newBlocks
+          properties:
+            datasetId:
+              type: string
+              format: did
+              description: Dataset identifier
+            knownHead:
+              type: string
+              format: multihash
+              description: Hash of the metadata head block known to the pusher
```

Comment:
What's the role of this during the push? Is it enough for server to simply look at the `newBlocks` to see what interval is being pushed to detect divergence/concurrency etc.?

Reply:
Agree, it's recoverable from the first new block in the batch.